### ISSUE TITLE
When clustered, out of sync clocks create records out of order

### ISSUE THREAD

**Issue by ajw625** [orbeon/orbeon-forms#3407]
We have Orbeon deployed in a cluster. There are two physical nodes where Orbeon is deployed. We ran into an issue when the clocks on these nodes got out of sync by about 20 seconds. Orbeon is populating the `LAST_MODIFIED_TIME` in the `ORBEON_FORM_DATA` table using a timestamp generated on the application server. This means that in this situation the order of the timestamps is different from the order the records were created. Since the latest record is retrieved based on the timestamp, the incorrect record may be retrieved.

It was decided that we should not rely on the the app server clocks being kept in sync and should instead use a timestamp retrieved from the database. Since we don't have direct control of the Orbeon source code, we have decided to resolve this by adding a database trigger:

    CREATE OR REPLACE TRIGGER orbeon_form_data_last_modified
             BEFORE INSERT ON orbeon_form_data
    FOR EACH ROW
    BEGIN
        :new.last_modified_time := systimestamp;
    END;
    /

We could have modified the existing trigger, but Oracle allows multiple triggers and this one is not likely to get replaced by a change in a future Orbeon release.

Another possible solution would be for the application not to set a value and simply use a default in the database, though I'm not sure if this would work with all the databases that Orbeon supports:

    alter table orbeon_form_data modify (
      last_modified_time default systimestamp
    );

I'm not sure how many others are clustering Orbeon but I thought you should know about this.

**Comment by avernet** [comment]
Yes, we shouldn't rely on the app server clock. My bad. Rookie mistake! We have a similar case in our implementation of form locking ([lease feature](http://blog.orbeon.com/2017/12/lease-lock.html)) coming out in 2017.2, and at least we haven't made the same mistake there. Besides, now that we have an index, we can just use it in all queries instead of relying on the `last_modified_time`, per #953. It will be more efficient, the queries will be simpler, and it will solve this problem.

But, I am keeping this open instead of marking it as duplicate of #953 as we should *also* compute the time on the database, if only to have more accurate timing information. I'm glad you have a workaround, but we should take care of this, and thank you @ajw625 for the detailed report.

**Comment by ajw625** [comment]
Sounds good. Thanks, Alex.
